"Archive Option" Feature

mattyg opened the pull request:
**Overview**
- Feature to allow a discord user with editor permission to archive a poll option, copying the UX of adding an option. 
- created an .env.example and basic developer documentation 
- move sql schema definition into `migrations` folder, to make their order explicit


**Deployment**
- Run migrations/1_archive_options.sql on database before deploying code changes

**WIP**
~~This currently introduces a bug where an VoteButton can display the wrong index under certain condition -- when there are 3+ VoteButtons and the 1st or 2nd is archived. I don't have the bandwidth to resolve it right now. I removed the index number from the VoteButton text as a workaround that's enough for my use case.~~ Edit: fixed

mattyg committed:
chore: add example .env

mattyg committed:
chore: order sql defintions as migrations

mattyg committed:
feat: sql migration to support archived options

mattyg committed:
feat: crud functions to archive an option

mattyg committed:
feat: controller functions to archive and option and retreive archived property

mattyg committed:
feat: discord frontend for archiving option

mattyg committed:
feat: remove index number from vote button ux

This is a workaround to avoid UX bug where buttons show wrong index after archiving a previous button.

mattyg committed:
feat: add basic developer documentation to docs/development-overview.md, link from README.md

mattyg committed:
fix: move developer docs into seperate file from README.md

mattyg committed:
fix: add missing semicolon

abrahammurciano commented on application/option.py:
```diff
@@ -65,6 +68,11 @@ def vote_count(self) -> int:
 		return len(self.__votes)
 
+	@property
+	def archived(self) -> str:
```
It's not a string, it's an Optional[datetime]

mattyg replied:
you're looking at an older commit -- already fixed

abrahammurciano commented on application/option.py:
```diff
@@ -65,6 +68,11 @@ def vote_count(self) -> int:
 
+	@property
+	def archived(self) -> str:
+		"""The archived date of the option, or null if not archived"""
```
```suggestion
		"""The archived date of the option, or None if not archived"""
```

mattyg replied:
fixed

abrahammurciano commented on application/option.py:
```diff
@@ -143,7 +158,7 @@ async def create_options(
 		"""
 		options = [
-			Option(None, label, (), poll, index, author_id)
+			Option(None, label, (), poll, index, author_id, False)
```
```suggestion
			Option(None, label, (), poll, index, author_id, None)
```

mattyg replied:
you're looking at an older commit -- already fixed

abrahammurciano commented:
Thanks for this! I'll review it sometime in the next few days hopefully

abrahammurciano commented on docs/development-overview.md:
```diff
@@ -0,0 +1,28 @@
+# Development Overview
+
+Requirements: Python, MySQL
```
I actually use posgresql. Can't guarantee that the syntax works on MySQL (though it might).

```suggestion
Requirements: Python, Posgresql
```

mattyg replied:
:+1:  fixed

abrahammurciano commented on docs/development-overview.md:
```diff
@@ -0,0 +1,28 @@
+
+### Run server
+- Activate venv
+  - `source bin/activate`
```
```suggestion
  - `source venv/bin/activate`
```

mattyg replied:
:+1: fixed

abrahammurciano commented on presentation/ui/archive_option_button.py:
```diff
@@ -0,0 +1,56 @@
+						poll, int(reply.content)-1, inter.message
+					)
+					await reply.add_reaction("✅")
+				except ValueError as e:
```
```suggestion
				except ValueError:
```

mattyg replied:
:+1: fixed

abrahammurciano commented on presentation/ui/poll_view.py:
```diff
@@ -33,6 +38,14 @@ def __add_add_option_button(self):
+		if (
+			not self.__poll.is_expired
+			and self.__poll.allowed_editors
+			and len(self.__poll.options) > 1
```
This check is not necessary. There can't be polls without options.

mattyg replied:
I included it to prevent users from archiving the *last* option in a poll.

mattyg replied:
But actually on second thought I think the user should be able to remove the last remaining option, as they can still add new ones.

Modified the check above to `len(self.__poll_options) > 0` to hide the button when there are *no* options available to archive.

abrahammurciano commented on presentation/ui/poll_view.py:
```diff
@@ -15,8 +16,12 @@ def __init__(self, bot: "Paul", poll: Poll):
+	def __create_view(self):
 		self.__add_vote_buttons()
 		self.__add_add_option_button()
+		self.__add_archive_option_button()
```
The addition of another button means the number of options limit is now 22 instead of 23. This should be updated in the command validation as well as in __add_add_option_button

mattyg replied:
:+1: fixed

abrahammurciano commented:
@mattyg What's the intention of this feature? Is it to delete an option so that it is no longer visible and no record of its existence is visible to the user? Or is it simply to disable voting on said option?

If it's the former, then the option and all its votes can be deleted from the database (database should delete votes automatically, I think cascade delete is enabled), and this feature can be called "remove option" rather than "archive". (Regarding option indices, I think the option index can be removed from the database. IIRC I added them to fix a bug that could have been, and subsequently was, fixed another way.

If it's the latter, then I would say that the options should all stay in the same list, and they should all be displayed, but some indication that it is archived (like the word "archived" in parentheses, and the option label ~~crossed out~~, as well as the vote button being grey) should be displayed instead of not displaying it at all.

I'm okay with either one being implemented (preferably not both because they achieve basically the same thing). So which one is it?

mattyg committed:
fix: make option.archived field timezone aware in postgres

mattyg committed:
fix: set and get option.archived as timezone aware

mattyg committed:
fix: revert splitting of options into __options & __options_archived in poll application. Instead filter getters to only archived or active

mattyg committed:
feat: display list of archived options in PollEmbed

mattyg committed:
fix: ensure option index label matches with expected input to 'What is the number of the option you want to archive?'

mattyg commented:
> @mattyg What's the intention of this feature? Is it to delete an option so that it is no longer visible and no record of its existence is visible to the user? Or is it simply to disable voting on said option?

Yes, good point. My intention was to display the list of archived options, underneath the list of "active" options.

I updated this PR to do so, and also fixed the bug I noted in the original comment. (heads up that I modified the migration `1_archive_options.pqsl`, so it will have to be re-run on top of the original db state)

I guess that while "functionally" we are archiving the option (i.e. it is not deleted in the db), from the user's perspective it might as well be removing the option, since they cannot un-archive it. It may make sense for the discord UX to use the language "Removed" rather than "Archived". Happy to just follow your preference on that.

mattyg committed:
chore: updtae migration file extension for consistency

abrahammurciano commented:
Very well. In that case I'd rather delete the option entirely from the database (and cascade delete the votes) as it would just be taking up unnecessary space. There's no need to display the deleted options. It would also make it a lot simpler to deal with option indices.

I would like to remove the index column from the option table (and class) entirely, and instead rely on the order that the options are returned by the query. If you'd like to do that, it will make it easier for you to delete options.

I'll finish reviewing your actual code soon, but in the meantime you can make it do a full proper delete instead of archiving.

abrahammurciano commented:
Also I'm not too keen on allowing the same people who can add options to delete options. Adding an option is less harmful than deleting one. If you have a poll and want to allow anyone to suggest more options, you don't necessarily want to allow anyone to remove other people's options.

Possible solutions might include:
- have a separate option called allowed_deleters or something (most flexible solution, but I'd rather keep the interface as simple as possible)
- only allow poll author and option author to delete an option (just like only poll author is allowed to close the poll)
- piggyback on discord's delete messages permission (might be worth doing the same for closing polls) to determine who can delete options
- combination of previous two solutions (my favourite solution)
- allow any editor to delete options, but only if (new) argument allow_deleting_options is set to true (default to false)

mattyg commented:
To be clear, for my use case, I *want* to display a list of the archived options. If that doesn't fit with your vision for the project please let me know.

This archive feature doesn't affect use of the index column. It does rely on the order that the options are returned by the query.

My preference for revising permissions is option 2, for simplicity. Honestly I don't the bandwidth to do any more significant refactoring. Please confirm if that's a reasonable compromise for you and I'll implement it.

abrahammurciano commented:
Fair enough. That is acceptable. In that case you may keep the wording as "archive" instead of "remove".

Thanks again for your contribution! Really appreciate it

abrahammurciano reviewed (commented):
I've not had the time to go over all of it yet, but to get you started, here are some observations on the portion that I have gone over so far.

mattyg committed:
fix: specify return value as optional in Option.archived()

mattyg committed:
fix: use language 'None' instead of 'null' in docstring

mattyg committed:
fix: specify postgres requirement in dev docs

mattyg committed:
fix: venv path in dev docs

mattyg committed:
fix: remove unused variable definition

mattyg committed:
fix: reduce max options from 23 to 22

mattyg committed:
fix: allow user to archive last remaining option in poll

mattyg committed:
feat: grant *only* the poll author permissions to archive options

mattyg commented:
Modified to only allow poll author to archive options.

mattyg committed:
fix: defer initial poll response message to prevent timeout

mattyg committed:
fix: pytz utc import property

mattyg committed:
fix: add psql query to ensure options_extended_view is recreated properly
